gae-to-gae-auth/client/client.py

```diff
@@ -0,0 +1,137 @@
+
+
+import argparse
+import json
+import logging
+import os
+import requests
+import sys
+import time
+
+from google.auth import crypt, jwt
+
+GOOGLE_ID_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token"
+
+def make_signed_jwt(service_account_key_path, audience, subject=None, ttl_seconds=3600, **kwargs):
+  """Generate a JWT token signed by a Google Cloud Platform service account.
+
+  Args:
+    service_account_key_path: (str) A full path to a Google Cloud Platform
+      service account file (in JSON format) to use to sign the JWT.
+    audience: (str) The recipients that the JWT is intended for.
+
+  Keyword arguments:
+    subject: (str) An optional claim that identifies the subject of the JWT.
+    ttl_seconds: (int) The time, in seconds, after which the JWT will not
+      be accepted.
+    **kwargs: Any additional keyword arguments are encoded as JSON in the
+      payload of the JWT.
+  """
+  with open(os.path.expanduser(service_account_key_path), mode='rt') as f:
+      service_account_info = json.load(f)
+
+  signer = crypt.RSASigner.from_service_account_info(service_account_info)
+
+  now = int(time.time())
+  if subject is None:
+    subject = service_account_info['client_email']
+  payload = {
+      'iat': now,
+      'exp': now + ttl_seconds,
+      'sub': subject,
+      'iss': service_account_info['client_email'],
+      'email': service_account_info['client_email'],
+      'aud': audience,
+  }
+  payload.update(**kwargs)
+  logging.debug("JWT payload: %r", payload)
+  signed_jwt = jwt.encode(signer, payload)
+  logging.debug("Signed JWT: %s", signed_jwt)
+
+  return signed_jwt
+
+
+def make_google_id_token(service_account_key_path, audience, **kwargs):
+  signed_jwt = make_signed_jwt(
+      service_account_key_path=service_account_key_path,
+      audience=GOOGLE_ID_AUDIENCE,
+      target_audience=audience,
+  )
+
+  # send the JWT to Google Token endpoints to request Google ID token
+
+  params = {
+      'grant_type': 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+      'assertion': signed_jwt
+  }
+  headers = {"Content-Type": "application/x-www-form-urlencoded"}
+  response = requests.post('https://www.googleapis.com/oauth2/v4/token', data=params, headers=headers)
+  res = response.json()
+
+  logging.debug("JSON response: %s", json.dumps(res, indent=2))
+
+  google_id_token = res['id_token']
+  logging.debug('Google ID token: %s', google_id_token)
+
+  claims = jwt.decode(google_id_token, verify=False)
+  logging.info("Google ID token claims: %s", json.dumps(claims, indent=2))
+  return google_id_token
+
+
+def run(args):
+  audience = '{project}.appspot.com'.format(
+    project=args.project_id)
+  key_path = args.service_account_path
+  if args.use_google_id_token:
+     bearer = make_google_id_token(
+       service_account_key_path=key_path,
+       audience=audience,
+   )
+  else:
+    bearer = make_signed_jwt(
+      service_account_key_path=key_path,
+      audience=audience,
+    )
```

Comment:
The `make_signed_jwt()` use-case can actually be achieved using `google.auth.jwt.Credentials` as follows:

```
creds = google.auth.jwt.Credentials.from_service_account_file("/path/to/service-account-key.json", audience=token_uri)
session = google.auth.transport.requests.AuthorizedSession(creds)
response = session.post(...)
```

Reply:
I'm still trying to find out if we can also construct an AuthorizedSession that relies on Google ID tokens.

Reply:
I switched the `make_signed_jwt` out for the `google.auth` library, and observe the following error:

```
Encountered unexpected error from ProtoRPC method implementation: UserNotFoundError ()
Traceback (most recent call last):
File "/base/alloc/tmpfs/dynamic_runtimes/python27/54c5883f70296ec8_unzipped/python27_lib/versions/third_party/protorpc-1.0/protorpc/wsgi/service.py", line 181, in protorpc_service_app
response = method(instance, request)
File "/base/data/home/apps/s~buser-gae-greeting/20180103t142338.406684610653599380/lib/endpoints/api_config.py", line 1298, in invoke_remote
return remote_method(service_instance, request)
File "/base/alloc/tmpfs/dynamic_runtimes/python27/54c5883f70296ec8_unzipped/python27_lib/versions/third_party/protorpc-1.0/protorpc/remote.py", line 414, in invoke_remote_method
response = method(service_instance, request)
File "/base/data/home/apps/s~buser-gae-greeting/20180103t142338.406684610653599380/main.py", line 61, in greet
user = endpoints.get_current_user()
File "/base/data/home/apps/s~buser-gae-greeting/20180103t142338.406684610653599380/lib/endpoints/users_id_token.py", line 119, in get_current_user
return users.User(user_info.email)
File "/base/alloc/tmpfs/dynamic_runtimes/python27/54c5883f70296ec8_unzipped/python27_lib/versions/1/google/appengine/api/users.py", line 122, in __init__
raise UserNotFoundError
UserNotFoundError
```

I believe this is because the Endpoints Frameworks module `users_id_token` looks for an email in the request (see (code)[https://github.com/cloudendpoints/endpoints-python/blob/6e3fd50b5bb42889ec3d41d66cd7fcf883af6f31/endpoints/users_id_token.py#L119]), but the `google.auth` library does not set the email field when using a service account.

Not sure what the correct fix is here. Any thoughts, @jonparrott ?

Reply:
I fixed this by additional `additional_claims={'email': '...'}`.

Reply:
Yes, additional claims is the way to go!
Thanks for updating this code.

Reply:
`users_id_token` requires an OpenID Connect ID Token (which is what the other branch creates). This branch creates a JWT bearer token which *can* identify a principal but doesn't necessarily have to.  You'll need to specify a different auth strategy to endpoint to use jwt bearer, see https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/endpoints/getting-started/openapi-appengine.yaml#L101

Reply:
Interesting, I'm curious how the endpoints framework deals with that -it's quite different outside of GAE standard.

Reply:
Jon : Can you expand on what's quite different? Maybe we're doing something wrong here?

Reply:
I think it's probably fine, it was just unexpected (to me) how you did this. :)